### DISCUSSION TITLE
GSOC 2022 : "Switching to Conscrypt as the default security provider"

### DISCUSSION THREAD

**Discussion by Umesh-01** [robolectric/robolectric#7199] [General]
Hello guys,

I am Umesh Singh, a pre-final year student pursuing Bachelor's in Computer Science and Engineering from Chandigarh University.
I am comfortable with Java and Cpp programming languages. I have a firm grasp of Data Structures and Algorithms and Object-Oriented Programming. 

I am also familiar with Android Studio and other tools. This project falls well in my current capabilities.

I would like to work on this project and want to get familiar with the codebase of Robolectric. So, can you tell me more about this and guide me further about the same. I am looking forward to hear from you and excited to work on this project.

**Comment by utzcoz** [comment]
Hi @Umesh-01, welcome to Robolectric. Please read http://robolectric.org/building-robolectric/ and http://robolectric.org/contributing/ to setup development environment on your machine firstly. It's recommended to try some [starter issues](https://github.com/robolectric/robolectric/issues?q=is%3Aissue+is%3Aopen+label%3A%22starter+issue%22) to learn what Robolectric is and how it works. The contributor applying period has started, and it's recommended to view issues related to this project(the idea list has context) to learn the current status of switching `Conscrypt` as default security provider, and prepare your proposal before applying. Feel free to ask any question about this project.

**Comment by Abdulrahman-Ragab-01** [comment]
Hi,
My name is Abdulrahman Ragab an undergraduate Software Engineer student

I am interested in this project..
I know the requirements of the project that I should switch Conscrypt as default security provider. 
I am aware with Java, C++, OOP, Datastructure and Algorithms, also I am aware of android studio and I had previously build many of android applications.
Could you please tell me more details and what task should I work on to help me dive deeper in the codebase?

**Reply by utzcoz** [reply]
@Abdulrahman1890  welcome to Robolectric. You can read https://github.com/robolectric/robolectric/discussions/7199#discussioncomment-2506700 firstly.

**Comment by hoisie** [comment]
This project is quite straightforward - currently Robolectric uses BouncyCastle as the JCE security provider. This means that whenever  Ciphers/KeyFactory/etc.. is used, BouncyCastle will get used.

In real Android, the default JCE provider is [Conscrypt](https://source.android.com/devices/architecture/modular-system/conscrypt). This is a Java wrapper over BoringSSL. The goal is to have Robolectric switch from BouncyCastle by default to Conscrypt.

We have had some pretty big issues with BouncyCastle in the past, most notably https://github.com/robolectric/robolectric/issues/5456. Also https://github.com/robolectric/robolectric/issues/6521 is annoying.

Part of this project will be to evaluate what kind of tests may break due to this change, and I can help with that.

**Comment by hoisie** [comment]
This is the relevant parts of the code where BC is attached:

https://github.com/robolectric/robolectric/blob/master/robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java#L149-L151

Also it's worth mentioning that Conscrypt is much faster than BC as it is optimized native code.
